src/eiffel_graphql_api/graphql/db/database.py

```diff
@@ -42,20 +58,98 @@ def connect(mock):
                                              urllib.parse.quote(password),
                                              host, port)
     else:
-        url = "mongodb://{}:{}".format(host, port),
+        url = "mongodb://{}:{}".format(host, port)
     CLIENT = mongo_client(url, replicaset=replicaset)
     DATABASE = CLIENT[database_name]
+    # 'server_info' will create a connection against the MongoDB effectively testing
+    # the connection for us. This means that the 'connect' method will block until
+    # a connection can be established or 'ServerSelectionTimeoutError' is raised.
+    LOGGER.info("%r", CLIENT.server_info())
+    LOGGER.info("Connected.")
+
+
+def retry(fixture, timeout, *args, **kwargs):
+    """Retry call to fixture for 'timeout' seconds.
+
+    :param fixture: Fixture to call upon.
+    :type fixture: :obj:`func`
+    :param timeout: How long to wait for fixture to succeed. In seconds.
+    :type timeout: float
+    :param args: Positional argument to pass onto fixture.
+    :type args: tuple
+    :param kwargs: Keyword arguments to pass onto fixture.
+    :type kwargs: dict
+    """
+    end = time.time() + timeout
+    while time.time() < end:
+        try:
+            fixture(*args, **kwargs)
+            break
+        except Exception as exception:  # pylint:disable=broad-except
+            LOGGER.warning("%r", exception)
+            LOGGER.warning("Retrying %r", fixture)
+            time.sleep(0.1)
+    else:
+        raise TimeoutError("Timeout after %rs calling fixture %r" % (timeout, fixture))
 
 
 def get_database(mock=False):
+    """Create and wait for a database connection.
+
+    :param mock: Whether the server should be mocked or not. Used for tests.
+    :type mock: bool
+    :return: Database instance.
+    :rtype: :obj:`pymongo.MongoClient.collection`
+    """
     global DATABASE
     if DATABASE is None:
-        connect(mock)
+        retry(connect, float(RETRY_TIMEOUT), mock)
     return DATABASE
 
 
 def get_client(mock=False):
+    """Create and wait for a client connection.
+
+    :param mock: Whether the server should be mocked or not. Used for tests.
+    :type mock: bool
+    :return: Client instance.
+    :rtype: :obj:`pymongo.MongoClient`
+    """
     global CLIENT
     if CLIENT is None:
-        connect(mock)
+        retry(connect, float(RETRY_TIMEOUT), mock)
     return CLIENT
+
+
+def insert_to_db(event, _=None):
+    """Eiffel callback for storing events in database.
+
+    :param event: Event to store into database.
+    :type event: :obj:`eiffellib.events.eiffel_base_event.EiffelBaseEvent`
+    :param context: In which context was the event sent.
+    :type context: str
+    :return: Whether or not database insertion succeeded.
+    :rtype: bool
+    """
+    global DATABASE
+    if DATABASE is None:
+        with LOCK:
+            DATABASE = get_database()
+        return False  # Requeue event if DATABASE is not ready.
```

Comment:
But doesn't get_database() effectively set DATABASE via connect? Also, once we're connected, shouldn't we continue executing this method rather than returning False?